content/testing.md

```diff
@@ -0,0 +1,683 @@
+
+To ensure the database is clean, the [`xolvio:cleaner`](https://atmospherejs.com/xolvio/cleaner) package is useful. You can use it to reset the database in a `beforeEach` block:
+
+```js
+import { resetDatabase } from 'meteor/xolvio:cleaner';
+
+describe('my module', () => {
+  beforeEach(() => {
+    resetDatabase();
+  });
+});
+```
+
+This technique will only work on the server. If you need to reset the database from a client test, you can use a method to do so:
+
+```js
+import { resetDatabase } from 'meteor/xolvio:cleaner';
+
+Meteor.methods({
+  'test.resetDatabase': () => resetDatabase();
+});
+
+describe('my module', done => {
+  beforeEach(() => {
+    // We need to wait until the method call is done before moving on, so we
+    // use Mocha's async mechanism (calling a done callback)
+    Meteor.call('test.resetDatabase', done);
+  });
+});
+```
+
+As we've placed the code above in a test file, it *will not* load in normal development or production mode (which would a bad thing!). If you create a Atmosphere package with a similar feature, you should mark it as `testOnly` and it will similarly only load in test mode.
+
+<h3 id="generating-test-data">Generating test data</h3>
+
+Often it's sensible to create a set of data to run your test against. You can use standard `insert()` calls against your collections to do this, but often it's easier to create *factories* which help encode random test data. A great package to use to do this is [`dburles:factory`](https://atmospherejs.com/dburles/factory).
+
+In the Todos example app, we define a factory to describe how to create a test todo item, using the [`faker`](https://www.npmjs.com/package/Faker) NPM package:
+
+```js
+import faker from 'faker';
+
+Factory.define('todo', Todos, {
+  listId: () => Factory.get('list'),
+  text: () => faker.lorem.sentence(),
+  createdAt: () => new Date(),
+});
+```
+
+To use the factory in a test, we simply call `Factory.create`:
+
+```js
+// This creates a todo and a list in the database and returns the todo.
+const todo = Factory.create('todo');
+
+// If we have a list already, we can pass in the id and avoid creating another:
+const list = Factory.create('list');
+const todoInList = Factory.create('todo', { listId: list._id });
+```
+
+<h3 id="mocking-the-database">Mocking the database</h3>
+
+As `Factory.create` directly inserts documents into the collection that's passed into the `Factory.define` function, it can be a problem to use it on the client. However there's a neat isolation trick that you can do to replace the server-backed `Todos` [client collection](collections.html#client-collections) with a mocked out [local collection](#collections.html#local-collections), that's encoded in the `stub-collections` package (currently a local package in the Todos example application).
```

Comment:
We should come up with how to call out these "fake" testing packages. Maybe explicitly say - they work, but we hope someone will write a better one, or something?